### PR TITLE
Dockerized Services

### PR DESCRIPTION
**Pull request by bcongdon**
Changing docker setup to have each service/client in its own Docker container.

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -1,21 +1,124 @@
...
+    restart: always
+
+  groot-desktop-frontend:
+    build: groot-desktop-frontend
+    depends_on:
+      - groot
+      - groot-events-service
+      - groot-users-service
+      - groot-groups-service
+      - groot-quotes-service
```
Comment (narendasan) [inline docker-compose.yml]: not entirely sure what depends_on does but are these the correct service dependencies?

File: docker-compose.yml [hunk 2]
```diff
@@ -1,21 +1,124 @@
...
+  groot-voz:
+    build: groot-voz
+    ports:
+      - "5652:5652"
+    depends_on:
+      - db
+      - groot
+    restart: always
+
+volumes:
```
Comment (narendasan) [inline docker-compose.yml]: Where do log files go? like the access_log also I would eventually like to move towards log files

File: docker-compose.yml [hunk 3]
```diff
@@ -1,21 +1,124 @@
...
-     - .:/groot-deploy
-   ports:
-     - "8000:8000"
-     - "5000:5000"
-   depends_on:
-     - db
+      - "3306:3306"
+    restart: always
+
+  groot:
```
Comment (narendasan) [inline docker-compose.yml]: This should probably be changed to groot-api-gateway

File: database_setup.sql [hunk 4]
```diff
@@ -1,6 +1,10 @@
 CREATE DATABASE IF NOT EXISTS groot_caffeine_service_dev;
```
Comment (narendasan) [inline database_setup.sql]: is there a separate setup for prod dbs?

Comment (bcongdon) [inline database_setup.sql]: I think we're currently running services with the `_dev` suffix in prod...

File: docker-compose.yml (hunk 2 above)
Comment (bcongdon) [inline docker-compose.yml]: I can make a volume for log files. We also can grab the logs of each service through [docker logs](https://docs.docker.com/engine/reference/commandline/logs/)

File: docker-compose.yml (hunk 1 above)
Comment (bcongdon) [inline docker-compose.yml]: depends_on just specifies the startup order of services

File: database_setup.sql (hunk 4 above)
Comment (sameetandpotatoes) [inline database_setup.sql]: Yeah I was just following a convention of a `_dev` suffix for development environments and just the service name with no suffix for prod (at least for ruby services).

File: docker-compose.yml (hunk 2 above)
Comment (bcongdon) [inline docker-compose.yml]: OK, this is setup now, assuming we use the config defaults I added to groot-api-gateway: https://github.com/acm-uiuc/groot/blob/4a1ed0def4570d09d9b8bd7839d58b74842d30fd/README.md#config-values

File: docker-compose.yml [hunk 5]
```diff
@@ -1,21 +1,121 @@
...
-     - db
+      - "3306:3306"
+    restart: always
+
+  groot-api-gateway:
+    build: groot-api-gateway
+    ports:
+      - "8000:8000"
+    volumes: 
+      - groot_api_gateway_volume:/var/groot/
```
Comment (narendasan) [inline docker-compose.yml]: does the /var/groot need to change as well?

**Comment by narendasan** [comment]
In this PR can you also include the readme updates for the preferred development and deployment instructions (since there are theoretically 3 options)?

**Comment by narendasan** [comment]
Also docker_settings_init.sh needs to be updated (both to match arbor and the new docker urls)

**Comment by bcongdon** [comment]
I'm inclined to make the singular docker container deployment method unsupported. It's just not a good solution.

This would make docker_settings_init obsolete.

**Comment by narendasan** [comment]
Sounds good

**Comment by narendasan** [comment]
Well actually it might be useful for setting up non docker setups if we want to continue to support that

**Comment by narendasan** [comment]
Perhaps that and the documentation surrounding non-docker development if that is going to be a secondary way of working should be moved to a wiki on this repo

**Pull request merged**